# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [37]:
# Import your libraries:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sn
import sklearn
from sklearn.model_selection import train_test_split

# Challenge 1 - Explore the Scikit-Learn Datasets

Before starting to work on our own datasets, let's first explore the datasets that are included in this Python library. These datasets have been cleaned and formatted for use in ML algorithms.

First, we will load the diabetes dataset. Do this in the cell below by importing the datasets and then loading the dataset  to the `diabetes` variable using the `load_diabetes()` function ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_diabetes.html)).

In [38]:
# Your code here:
from sklearn.datasets import load_diabetes

diabetes = load_diabetes()


Let's explore this variable by looking at the different attributes (keys) of `diabetes`. Note that the `load_diabetes` function does not return dataframes. It returns you a Python dictionary.

In [39]:
# Your code here:
diabetes.keys()


dict_keys(['data', 'target', 'frame', 'DESCR', 'feature_names', 'data_filename', 'target_filename', 'data_module'])

#### The next step is to read the description of the dataset. 

Print the description in the cell below using the `DESCR` attribute of the `diabetes` variable. Read the data description carefully to fully understand what each column represents.

*Hint: If your output is ill-formatted by displaying linebreaks as `\n`, it means you are not using the `print` function.*

In [40]:
# Your code here:
print(diabetes["DESCR"])

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

  :Number of Instances: 442

  :Number of Attributes: First 10 columns are numeric predictive values

  :Target: Column 11 is a quantitative measure of disease progression one year after baseline

  :Attribute Information:
      - age     age in years
      - sex
      - bmi     body mass index
      - bp      average blood pressure
      - s1      tc, total serum cholesterol
      - s2      ldl, low-density lipoproteins
      - s3      hdl, high-density lipoproteins
      - s4      tch, total cholesterol / HDL
      - s5      ltg, possibly log of serum triglycerides level
      - s6      glu, blood sugar level

Note: Each of these 1

#### Based on the data description, answer the following questions:

1. How many attributes are there in the data? What do they mean?

1. What is the relation between `diabetes['data']` and `diabetes['target']`?

1. How many records are there in the data?

In [41]:
# Enter your answer here:
#1. How many attributes are there in the data? What do they mean?
    # There are 10 attributes in the data
    
#2. What is the relation between diabetes['data'] and diabetes['target']?

    # The diabetes['data'] contains the values of the 10 attributes mentioned above for each patient. 
    # It serves as the set of predictor variables or features. 
    #  On the other hand, diabetes['target'] contains the quantitative measure of disease progression one year after the baseline for each patient. It represents the response or outcome variable. So, when you train a machine learning model on this dataset, you'll be using diabetes['data'] as your input features to predict the corresponding diabetes['target'] values.

#3. How many records are there in the data?
    # There are 442 records or instances in the data.

#### Now explore what are contained in the *data* portion as well as the *target* portion of `diabetes`. 

Scikit-learn typically takes in 2D numpy arrays as input (though pandas dataframes are also accepted). Inspect the shape of `data` and `target`. Confirm they are consistent with the data description.

In [42]:
# Your code here:
print(diabetes["data"].shape)
print(diabetes["target"].shape)

(442, 10)
(442,)


# Challenge 2 - Perform Supervised Learning on the Dataset

The data have already been split to predictor (*data*) and response (*target*) variables. Given this information, we'll apply what we have previously learned about linear regression and apply the algorithm to the diabetes dataset.

#### Let's briefly revisit the linear regression formula:

```
y = β0 + β1X1 + β2X2 + ... + βnXn + ϵ
```

...where:

- X1-Xn: data 
- β0: intercept 
- β1-βn: coefficients 
- ϵ: error (cannot explained by model)
- y: target

Also take a look at the `sklearn.linear_model.LinearRegression` [documentation](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html).

#### In the cell below, import the `linear_model` class from `sklearn`. 

In [43]:
# Your code here:
from sklearn.linear_model import LinearRegression

#### Create a new instance of the linear regression model and assign the new instance to the variable `diabetes_model`.

In [44]:
# Your code here:
diabetes_model = LinearRegression()

#### Next, let's split the training and test data.

Define `diabetes_data_train`, `diabetes_target_train`, `diabetes_data_test`, and `diabetes_target_test`. Use the last 20 records for the test data and the rest for the training data.

In [45]:
# Your code here:
features = diabetes["data"]
target = diabetes["target"]

diabetes_data_train, diabetes_data_test = features[:-20], features[-20:]
diabetes_target_train, diabetes_target_test = target[:-20], target[-20:]

Fit the training data and target to `diabetes_model`. Print the *intercept* and *coefficients* of the model.

In [46]:
# Your code here:
diabetes_model.fit(diabetes_data_train, diabetes_target_train)

print(f"Intercept: {diabetes_model.intercept_}")
print(f"Coefficients: {diabetes_model.coef_}")

Intercept: 152.76429169049118
Coefficients: [ 3.06094248e-01 -2.37635570e+02  5.10538048e+02  3.27729878e+02
 -8.14111926e+02  4.92799595e+02  1.02841240e+02  1.84603496e+02
  7.43509388e+02  7.60966464e+01]


#### Inspecting the results

From the outputs you should have seen:

- The intercept is a float number.
- The coefficients are an array containing 10 float numbers.

This is the linear regression model fitted to your training dataset.

#### Using your fitted linear regression model, predict the *y* of `diabetes_data_test`.

In [47]:
# Your code here:
diabetes_model.predict(diabetes_data_test)


array([197.61898486, 155.44031962, 172.88875144, 111.53270645,
       164.79397301, 131.06765869, 259.12441219, 100.47873746,
       117.06005372, 124.30261597, 218.36868146,  61.19581944,
       132.24837933, 120.33293546,  52.54513009, 194.03746764,
       102.5756431 , 123.56778709, 211.03465323,  52.60221696])

#### Print your `diabetes_target_test` and compare with the prediction. 

In [48]:
# Your code here:
display(diabetes_target_test)

array([233.,  91., 111., 152., 120.,  67., 310.,  94., 183.,  66., 173.,
        72.,  49.,  64.,  48., 178., 104., 132., 220.,  57.])

#### Is `diabetes_target_test` exactly the same as the model prediction? Explain.

In [49]:
# Your explanation here:

# No, the diabetes_target_test values are not exactly the same as the model predictions. This discrepancy is expected for a few reasons:
# Linear Regression: It is an approximation that tries to fit a linear relationship to the data. 
# It's rare for predictions to be exactly accurate, especially with real-world data. 
# The idea is to get close enough to make meaningful predictions.

# Nature of Real-World Data: Real-world data often has noise, unaccounted variables, or non-linear relationships that a simple linear regression might not capture completely.

# Training and Testing Split: The model was trained on one subset of data and tested on another. 
# If the test set has some characteristics or patterns that weren't as prevalent in the training set, then the predictions might not be as accurate.

# While the predictions of the linear regression model might be close to the actual values, they are rarely, if ever, exactly the same. It's always good to evaluate the model's performance using various metrics (like Mean Absolute Error, Mean Squared Error, etc.) to understand how well it's doing.

# Bonus Challenge 1 - Hypothesis Testing with `statsmodels`

After generating the linear regression model from the dataset, you probably wonder: then what? What is the statistical way to know if my model is reliable or not?

Good question. We'll discuss that using Scikit-Learn in Challenge 5. But for now, let's use a fool-proof way by using the ([Linear Regression class of StatsModels](https://www.statsmodels.org/dev/regression.html)) which can also conduct linear regression analysis plus much more such as calcuating the F-score of the linear model as well as the standard errors and t-scores for each coefficient. The F-score and t-scores will tell you whether you can trust your linear model.

To understand the statistical meaning of conducting hypothesis testing (e.g. F-test, t-test) for slopes, read [this webpage](https://onlinecourses.science.psu.edu/stat501/node/297/) at your leisure time. We'll give you a brief overview next.

* The F-test of your linear model is to verify whether at least one of your coefficients is significantly different from zero. Translating that into the *null hypothesis* and *alternative hypothesis*, that is:

    ```
    H0 : β1 = β2 = ... = β10 = 0
    HA : At least one βj ≠ 0 (for j = 1, 2, ..., 10)
    ```

* The t-tests on each coefficient is to check whether the confidence interval for the variable contains zero. If the confidence interval contains zero, it means the null hypothesis for that variable is not rejected. In other words, this particular vaiable is not contributing to your linear model and you can remove it from your formula.

Read the documentations of [StatsModels Linear Regression](https://www.statsmodels.org/dev/regression.html) as well as its [`OLS` class](https://www.statsmodels.org/dev/generated/statsmodels.regression.linear_model.OLS.html) which stands for *ordinary least squares*.

#### In the next cell, analyze `diabetes_data_train` and `diabetes_target_train` with the linear regression model of `statsmodels`. Print the fit summary.

Your output should look like:

![statsmodels regression](../statsmodels.png)

In [50]:
# Your code here:
import statsmodels.api as sm

# Add a constant to the predictor variables
diabetes_data_train_with_const = sm.add_constant(diabetes_data_train)

# Fit the model using OLS (Ordinary Least Squares)
model = sm.OLS(diabetes_target_train, diabetes_data_train_with_const).fit()

# Print the summary
print(model.summary())


                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.512
Model:                            OLS   Adj. R-squared:                  0.500
Method:                 Least Squares   F-statistic:                     43.16
Date:                Sun, 27 Aug 2023   Prob (F-statistic):           4.65e-58
Time:                        11:13:57   Log-Likelihood:                -2281.1
No. Observations:                 422   AIC:                             4584.
Df Residuals:                     411   BIC:                             4629.
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        152.7643      2.658     57.468      0.0

#### Interpreting hypothesis testing results

Answer the following questions in the cell below:

1. What is the F-score of your linear model and is the null hypothesis rejected?
    - The F-statistic (or F-score) for the linear model is 43.16. The associated p-value for this F-statistic is 4.65e-58(verylow, close to 0). If the p is low is considered evidence to reject the null hypoth in favor of the alterative hypothesis. Given the very small p-value we reject the null hypothesis. This means that at least one of the predictors is useful for predicting the target.

2. Does any of the t-tests of the coefficients produce a confidence interval containing zero? What are they?
    -Yes, several t-tests of the coefficients have a confidence interval that contains zero. They are:
    
    -x1: The 95% confidence interval is [-120.167, 120.779]
    -x5: The 95% confidence interval is [-1647.669, 19.445]
    -x6: The 95% confidence interval is [-183.857, 1169.457]
    -x7: The 95% confidence interval is [-328.566, 534.248]
    - x8: The 95% confidence interval is [-144.338, 513.545]
    - x10: The 95% confidence interval is [-58.151, 210.344]

For these coefficients, the confidence intervals include zero, which suggests that these predictors may not be statistically significant in the model.

3. How will you modify your linear reguression model according to the test results above?

Based on the test results:

Since the p-values for coefficients x1, x5 to x8, and x10 are relatively high and their confidence intervals contain zero, these predictors are candidates for removal from the model. They might not be contributing meaningfully to the prediction of the target.

We should consider a simplified model that excludes these less significant predictors. By doing so, we can achieve a model that is more interpretable and possibly better generalized to new data.

It's also essential to consider the context and domain knowledge. Sometimes a variable, even if not statistically significant in a particular dataset, might be of clinical or scientific importance and should be retained.

After modifying the model by dropping insignificant predictors, it would be good to refit the model and recheck its performance metrics to ensure that the model's predictive power hasn't diminished.

In [51]:
# Your answers here:

# Challenge 3 - Peform Supervised Learning on a Pandas Dataframe

Now that we have dealt with data that has been formatted for scikit-learn, let's look at data that we will need to format ourselves.

In the next cell, load the `auto-mpg.csv` file included in this folder and assign it to a variable called `auto`.

In [52]:
# Your code here:

auto = pd.read_csv("auto-mpg.csv")


Look at the first 5 rows using the `head()` function:

In [53]:
# Your code here:
auto.head(5)


mpg  cylinders  displacement  horse_power  weight  acceleration  \
0  18.0          8         307.0        130.0    3504          12.0   
1  15.0          8         350.0        165.0    3693          11.5   
2  18.0          8         318.0        150.0    3436          11.0   
3  16.0          8         304.0        150.0    3433          12.0   
4  17.0          8         302.0        140.0    3449          10.5   

   model_year                       car_name  
0          70  \t"chevrolet chevelle malibu"  
1          70          \t"buick skylark 320"  
2          70         \t"plymouth satellite"  
3          70              \t"amc rebel sst"  
4          70                \t"ford torino"

Evaluate the data to ensure that all numeric columns are correctly detected as such by pandas. If a column is misclassified as object, coerce it to numeric.

In [54]:
# Your code here:
auto.dtypes


mpg             float64
cylinders         int64
displacement    float64
horse_power     float64
weight            int64
acceleration    float64
model_year        int64
car_name         object
dtype: object

What is the newest model year and the oldest model year?

In [55]:
# Your code here:

oldest_model_year = auto['model_year'].min()
newest_model_year = auto['model_year'].max()

print(f"The oldest model year is: {oldest_model_year}")
print(f"The newest model year is: {newest_model_year}")

The oldest model year is: 70
The newest model year is: 82


Check the dataset for missing values and remove all rows containing at least one missing value.

In [56]:
# Your code here:
auto.isnull().sum()

mpg             0
cylinders       0
displacement    0
horse_power     6
weight          0
acceleration    0
model_year      0
car_name        0
dtype: int64

In [57]:
# Your code here:
auto.drop("horse_power", inplace = True, axis = 1)

Find the frequency table for the `cylinders` column using the `value_counts()` function. How many possible values of cylinders are there?

In [58]:
frequency_table = auto["cylinders"].value_counts()
frequency_table

4    204
8    103
6     84
3      4
5      3
Name: cylinders, dtype: int64

We would like to generate a linear regression model that will predict mpg. To do this, first drop the `car_name` column since it does not contain any quantitative data. Next separate the dataframe to predictor and response variables. Separate those into test and training data with 80% of the data in the training set and the remainder in the test set. 

Assign the predictor and response training data to `X_train` and `y_train` respectively. Similarly, assign the predictor and response test data to `X_test` and `y_test`.

*Hint: To separate data for training and test, use the `train_test_split` method we used in previous labs.*

In [59]:
# Your code here:
#1. Drop car_name
auto.drop("car_name", inplace = True, axis = 1)

In [60]:
#2. Separate Df to predictor and response variables

features = auto.drop("mpg", axis = 1)
target = auto["mpg"]

X_train, X_test, y_train, y_test = train_test_split(features, target, test_size = 0.20)

Now we will processed and peform linear regression on this data to predict the mpg for each vehicle. 

#### In the next cell, create an instance of the linear regression model and call it `auto_model`. Fit `auto_model` with your training data.

In [61]:
# Your code here:
auto_model = LinearRegression()

auto_model.fit(X_train, y_train)
auto_model.score(X_test, y_test)


0.816703194579669

# Challenge 4 - Evaluate the Model

In addition to evaluating your model with F-test and t-test, you can also use the *Coefficient of Determination* (a.k.a. *r squared score*). This method does not simply tell *yes* or *no* about the model fit but instead indicates how much variation can be explained by the model. Based on the r squared score, you can decide whether to improve your model in order to obtain a better fit.

You can learn about the r squared score [here](https://en.wikipedia.org/wiki/Coefficient_of_determination). In the end, we want the r-squared score to be as high as possible.

#### In the next cell, compute the predicted *y* based on `X_train` and call it `y_pred`. Then calcualte the r squared score between `y_pred` and `y_train` which indicates how well the estimated regression model fits the training data.

*Hint: r squared score can be calculated using `sklearn.metrics.r2_score` ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html)).*

In [62]:
# Your code here:
from sklearn.metrics import r2_score

y_pred = auto_model.predict(X_train)
r2_score(y_train, y_pred)

0.8058271407499401

#### Our next step is to evaluate the model using the test data. 

We would like to ensure that our model is not overfitting the data. This means that our model was made to fit too closely to the training data by being overly complex. If a model is overfitted, it is not generalizable to data outside the training data. In that case, we need to reduce the complexity of the model by removing certain features (variables).

In the cell below, use the model to generate the predicted values for the test data and assign them to `y_test_pred`. Compute the r squared score of the predicted `y_test_pred` and the oberserved `y_test` data.

In [63]:
# Your code here:
y_test_pred = auto_model.predict(X_test)
r2_score(y_test, y_test_pred)

0.816703194579669

#### Explaining the results

The r squared scores of the training data and the test data are pretty close (0.8146 vs 0.7818). This means our model is not overfitted. However, there is still room to improve the model fit. Move on to the next challenge.

# Challenge 5 - Improve the Model Fit

While the most common way to improve the fit of a model is by using [regularization](https://datanice.github.io/machine-learning-101-what-is-regularization-interactive.html), there are other simpler ways to improve model fit. The first is to create a simpler model. The second is to increase the train sample size.

Let us start with the easier option and increase our train sample size to 90% of the data. Create a new test train split and name the new predictors and response variables `X_train09`, `X_test09`, `y_train09`, `y_test09`.

In [64]:
# Your code here:
X_train09, X_test09, y_train09, y_test09 = train_test_split(features, target, test_size = 0.10)

Initialize a new linear regression model. Name this model `auto_model09`. Fit the model to the new sample (training) data.

In [65]:
# Your code here:
auto_model09 = LinearRegression()
auto_model09.fit(X_train09,y_train09)

LinearRegression()

Compute the predicted values and r squared score for our new model and new sample data.

In [66]:
# Your code here:
y_pred09 = auto_model09.predict(X_train09)
r2_score(y_train09, y_pred09)

0.8059023822911883

Compute the r squared score for the smaller test set. Is there an improvement in the test r squared?

In [67]:
# Your code here:
y_test_pred09 = auto_model09.predict(X_test09)
r2_score(y_test09, y_test_pred09)

0.8254521875467091

In [ ]:
#there was an improvement about 2%

# Bonus Challenge 2 - Backward Elimination 

The main way to produce a simpler linear regression model is to reduce the number of variables used in the model. In scikit-learn, we can do this by using recursive feature elimination. You can read more about RFE [here](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.RFE.html).

In the next cell, we will import RFE

In [68]:
from sklearn.feature_selection import RFE

Follow the documentation and initialize an RFE model using the `auto_model` linear regression model. Set `n_features_to_select=3`

In [69]:
# Your code here:


Fit the model and print the ranking

In [70]:
# Your code here:


Feature importance is ranked from most important (1) to least important (4). Generate a model with the three most important features. The features correspond to variable names. For example, feature 1 is `cylinders` and feature 2 is `displacement`.

Perform a test-train split on this reduced column data and call the split data `X_train_reduced`, `X_test_reduced`, `y_test_reduced`, `y_train_reduced`. Use an 80% split.

In [71]:
# Your code here:


Generate a new model called `auto_model_reduced` and fit this model. Then proceed to compute the r squared score for the model. Did this cause an improvement in the r squared score?

In [72]:
# Your code here: 


#### Conclusion

You may obtain the impression from this lab that without knowing statistical methods in depth, it is difficult to make major progress in machine learning. That is correct. If you are motivated to become a data scientist, statistics is the subject you must be proficient in and there is no shortcut. 

Completing these labs is not likely to make you a data scientist. But you will have a good sense about what are there in machine learning and what are good for you. In your future career, you can choose one of the three tracks:

* Data scientists who need to be proficient in statistical methods.

* Data engineers who need to be good at programming.

* Data integration specialists who are business or content experts but also understand data and programming. This cross-disciplinary track brings together data, technology, and business and will be in high demands in the next decade.